google/cloud/storage/internal/grpc_client.cc

```diff
@@ -129,39 +120,34 @@ Status TimeoutError(std::chrono::milliseconds timeout, std::string const& op) {
                     "] while waiting for " + op);
 }
 
-struct WaitForFinish {
-  std::unique_ptr<GrpcClient::WriteObjectStream> stream;
-  void operator()(future<StatusOr<google::storage::v2::WriteObjectResponse>>) {}
-};
-
-struct WaitForIdle {
-  std::unique_ptr<GrpcClient::WriteObjectStream> stream;
-  void operator()(future<bool>) {
-    auto finish = stream->Finish();
-    (void)finish.then(WaitForFinish{std::move(stream)});
-  }
-};
-
 StatusOr<QueryResumableUploadResponse> CloseWriteObjectStream(
     std::chrono::milliseconds timeout,
+    std::function<future<bool>()> const& create_watchdog,
     std::unique_ptr<GrpcClient::WriteObjectStream> writer,
     bool sent_last_message, google::cloud::Options const& options) {
   if (!writer) return TimeoutError(timeout, "Write()");
   if (!sent_last_message) {
-    auto pending = writer->WritesDone();
-    if (pending.wait_for(timeout) == std::future_status::timeout) {
+    auto watchdog = create_watchdog().then([&writer](auto f) {
+      if (!f.get()) return false;
       writer->Cancel();
-      pending.then(WaitForIdle{std::move(writer)});
+      return true;
+    });
+    (void)writer->Write(google::storage::v2::WriteObjectRequest{},
+                        grpc::WriteOptions().set_last_message());
+    watchdog.cancel();
+    if (watchdog.get()) {
+      writer->Close();
```

Comment:
do we need another watchdog that might cancel this `Close()` operation? we use one a few lines below here.

Reply:
No.  If we get to this point the streaming RPC is already canceled.  Setting up another watchdog, just to cancel the request again seems unlikely to speed up things.